### PR TITLE
Adding TabViewItems to a TabView via code

### LINKED ISSUE #559: [Bug] Adding tabs to the TabViewItems after constructing it does not update Carousel
### Description

If you make a TabView that uses the TabItems property, not the TabItemsSource, you cannot add tabs at runtime and then navigate to them.

The tab strip updates, but the CarouselView does not add the new TabItem Content to its ItemsSource at any time.

### Steps to Reproduce

1. Make a page with TabItems
2. Add an item using a button on the page
3. Click the tab in the strip, see it not load the Content you expected.

### Expected Behavior
Tabs can be added at runtime to the TabItems
### Actual Behavior
Tabs are ignored at runtime, and only produce a tab strip item.
### Basic Information

- Version with issue: Commit 43b8dbde
- Last known good version: Never
- IDE: VS 2019 16.7.7
- Platform Target Frameworks: Not relevant, Toolkit Forms code problem
  - iOS:  
  - Android:  
  - UWP:   
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

### Workaround

Use TabItemsSource, as there are listeners that actually do real things on it. The code for UpdateTabItems doesn't touch the carousel.

### Reproduction imagery

07073835-2B95-4C29-805F-684BC9A31E62

### Reproduction Link

[https://github.com/adammeaney/XamarinCommunityToolkit/tree/DynamicTabViewItems](https://github.com/adammeaney/XamarinCommunityToolkit/tree/DynamicTabViewItems)

### LINKED ISSUE #1219: [Bug] TabView Can't Be Added Or Removed After Created
### Steps to Reproduce

Remove TabViewItem:
1. Create TabView in XAML with many TabViews (example TabViewItem1, TabViewItem2, TabViewItem3, TabViewItem4, TabViewItem5)
2. Now when in `OnAppearing` try remove TabViewItem. Example `TabView.TabViewItems.RemoveAt(2)`. It will remove `TabViewItem3`
3. Now tab in the middle will be gone. Try select the tab next to it which is `TabViewItem4`

Add TabViewItem:
1. Create TabView in XAML with many TabViews (example TabViewItem1, TabViewItem2, TabViewItem3, TabViewItem4, TabViewItem5)
2. Now when in `OnAppearing` try add TabViewItem. Example 
```cs
TabView.TabViewItems.Add(new TabViewItem
{
    Text="Tab 6",
    Content = new ContentView()
});
```
 It will add Tab 6 to the most right tabs.
3. Now try click Tab 6 which is the one tab that we just added

### Expected Behavior
- For remove TabViewItem, the tabs should be resized after removed and clicking tab next to it not throwing 
- For add TabViewItem, the tabs should be able to get clicked after added

### Actual Behavior
- For remove TabViewItem, it will throw exception `System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.` I check `TabView.SelectedIndex` is set to 2 which should be correct and `TabView.TabItems` still at 4 which also correct
- For add TabViewItem, it will throw exception `System.IndexOutOfRangeException: Can't set CarouselView to position 5. ItemsSource has 5 items.`
### Basic Information

- Version with issue: v1.1.0
- Last known good version: -
- Platform Target Frameworks: 
  - iOS:  Not tested
  - Android: 11.0
  - UWP:  Not Tested

### Reproduction imagery

Remove TabViewItem :
image

### Reproduction Link

Can use this https://github.com/albilaga/XamarinIssues/tree/main/XamarinIssues/TabViews as base and then in here https://github.com/albilaga/XamarinIssues/blob/main/XamarinIssues/TabViews/MainTabPage.xaml.cs can add `OnAppearing` to remove or add `TabViewItem`.

### LINKED ISSUE #1412: [Bug] TabView - Adding tabs during OnAppearing content won't show
### Description
Tab content won't appear when the tabs are added in code behind in the OnAppearing method.

### Steps to Reproduce

1. Start a new Xamarin Forms project
2. Add a new TabView on a page
3. Give the TabView a x:Name property with a value of "MainTabs"
4. Go to the code behind of the page and override the OnAppearing method
5. Add a couple of tabs via this.MainTabs.TabItems.Add();
6. Run the app and see that the content part is not available.

### Expected Behavior
Content of tabs should appear on screen

### Actual Behavior
Tabs show up but the content will not show

### Workaround
Moving initialization of the tabs from OnAppearing to the constructor

### Reproduction Link
[Link to sample](https://github.com/GuidoNeele/TabViewBug)

```csharp
  public partial class MainPage : ContentPage
  {
    public MainPage()
    {
      InitializeComponent();

      //InitTabs(); // Works
    }

    protected override void OnAppearing()
    {
      InitTabs(); // Doesn't work
    }

    private void InitTabs()
    {
      ...

      this.MainTabs.TabItems.Add(new TabViewItem { Icon = smileIcon, Text = "Tab 1", Content = new Tab1LazyView() });
      this.MainTabs.TabItems.Add(new TabViewItem { Icon = bookIcon, Text = "Tab 2", Content = new Tab2LazyView() });
      this.MainTabs.TabItems.Add(new TabViewItem { Icon = bookIcon, Text = "Tab 3", Content = new View3() });
    }
  }
```

image

### PR DESCRIPTION
**Pull request by GuidoNeele**
### Description of Change ###

When changing the TabItems of the TabView the underlying CarouselView wasn't updated. This resulted in out of bounds exceptions if new tabs were added to an already existing filled TabView or resulted in not seeing the content when the tabs were added to an empty TabView. There also was an issue with removing tabs that would leave an empty space in the tab strip. That issue has also been fixed with this PR.

This PR is somewhat of a duplicate of PR https://github.com/xamarin/XamarinCommunityToolkit/pull/592 by @adammeaney. I stumbled upon the PR after I had created the bug fix and was looking which issues it solved. After I saw the PR I then updated the demo page to resemble that of Adams because this was requested in the conversation. I can also submit my changes to Adams PR but since that PR was stale for months I thought it'd be easier to submit mine that's all up to date with the main branche. 

### Bugs Fixed ###

- Fixes #1412 
- Fixes #1219 
- Fixes #559 

### API Changes ###

None

### Behavioral Changes ###

None

### PR Checklist ###

- [x] Has a linked Issue, and the Issue has been `approved`
- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard

- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

**Comment by GuidoNeele** [comment]
There is however one issue with this PR. Adding, inserting, moving and removing tabs all works. The tab strip is updated nicely, there's no overlapping or whitespaces left, but there is an issue with navigating between the tabs that I can't seem to solve. When clicking through the tabs the content part stops updating. For instance you have 5 tabs (tab1, tab2, tab3, tab4 and tab5), you navigate to tab3, the contents of tab3 will appear, you then click tab2 and the content of tab3 will stay on screen. You then have to click tab3 first again before you can move to another tab. It looks like CarouselView and the TabStrip aren't completely in sync but I haven't got the faintest idea what's going one.

**Comment by adammeaney** [comment]
I experienced similar things occasionally, which is one of the reasons I put off trying to finish that PR, as I just didn't have the time to look into it.

I am perfectly fine dropping my old one as one already up to date sounds nice.
Just let me know.

**Comment by GuidoNeele** [comment]
Anybody have an idea why this behavior is happening?
[Youtube video of strange effect in TabView after adding new tabs](https://youtu.be/gcmFSZe2I3Y)

@jsuarezruiz perhaps?

**Comment by GuidoNeele** [comment]
Any news on this?

File: samples/XCT.Sample/Pages/Views/TabView/DynamicTabsPage.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.ObjectModel;
+using Xamarin.CommunityToolkit.UI.Views;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Sample.Pages.Views.TabView
+{
+		public partial class DynamicTabsPage
+	{
+		private int TabCounter { get; set; } = 3;
```
Comment (nichojo89) [inline samples/XCT.Sample/Pages/Views/TabView/DynamicTabsPage.xaml.cs:10]: Should this private field start with underscore?
```suggestion
		private int _tabCounter { get; set; } = 3;
```

**Review by nichojo89 — CHANGES_REQUESTED** [review]

**Comment by jfversluis** [comment]
Now that we're so close to the sunsetting of Xamarin (and therefore also Xamarin Community Toolkit) unfortunately we won't be able to take this in anymore, we're really sorry about that. Nevertheless, thank you so much for your time and effort that you have put into this PR, we really appreciate it.

Please have a look at the evolution of the Xamarin Community Toolkit, [The .NET MAUI Community Toolkit](https://github.com/CommunityToolkit/Maui). A lot of development has been going on there. Hopefully this issue was already fixed in that codebase. If not, feel free to port this over to there.

Again, thank you so much for being a contributor and Xamarin user!

**Pull request closed without merging**